local/compose.go

```diff
@@ -136,14 +137,135 @@ func (s *composeService) Push(ctx context.Context, project *types.Project) error
 				if jm.Error != nil {
 					return errors.New(jm.Error.Message)
 				}
-				toProgressEvent(service.Name, jm, w)
+				toProgressEvent("Pushing "+service.Name, jm, w)
 			}
 			return nil
 		})
 	}
 	return eg.Wait()
 }
 
+func (s *composeService) Pull(ctx context.Context, project *types.Project) error {
+	configFile, err := cliconfig.Load(config.Dir(ctx))
+	if err != nil {
+		return err
+	}
+	info, err := s.apiClient.Info(ctx)
+	if err != nil {
+		return err
+	}
+
+	if info.IndexServerAddress == "" {
+		info.IndexServerAddress = registry.IndexServer
+	}
+
+	w := progress.ContextWriter(ctx)
+	eg, ctx := errgroup.WithContext(ctx)
+
+	for _, service := range project.Services {
+		service := service
```

Comment:
```suggestion
```

Reply:
It's needed because the body of the for loop is a goroutine, this line creates a new binding to the service variable so that each goroutine gets it's own `service`, without this all the goroutines would get the same `service`

Reply:
nit: in these cases I try to set a different variable name, that makes it a bit more obvious why we need a separate variable, rather than shadow the variable name.

Reply:
Legit, will change